packages/@sanity/types/src/validation/types.ts

```diff
@@ -1,9 +1,289 @@
+  | {flag: 'precision'; constraint: number}
+  | {flag: 'lessThan'; constraint: number}
+  | {flag: 'greaterThan'; constraint: number}
+  | {flag: 'stringCasing'; constraint: 'uppercase' | 'lowercase'}
+  | {flag: 'block'; constraint: BlockValidator}
+  | {flag: 'assetRequired'; constraint: {assetType: 'Asset' | 'Image' | 'File'}}
+  | {
+      flag: 'regex'
+      constraint: {
+        pattern: RegExp
+        name?: string
+        invert: boolean
+      }
+    }
+  | {
+      flag: 'uri'
+      constraint: {
+        options: {
+          scheme: RegExp[]
+          allowRelative: boolean
+          relativeOnly: boolean
+          allowCredentials: boolean
+        }
+      }
+    }
+
+// this is used to get allow index access (e.g. `RuleSpec['constraint']`) to
+// constraint when a rule spec might not have a 'constraint` prop
+type ConditionalIndexAccess<T, U> = U extends keyof T ? T[U] : undefined
+
+export type RuleSpecConstraint<T extends RuleSpec['flag']> = ConditionalIndexAccess<
+  Extract<RuleSpec, {flag: T}>,
+  'constraint'
+>
+
+/**
+ * A context object passed around during validation. This includes the
+ * `Rule.custom` context.
+ *
+ * e.g.
+ *
+ * ```js
+ * Rule.custom((_, validationContext) => {
+ *   // ...
+ * })`
+ * ```
+ */
+export type ValidationContext = {
+  parent?: unknown
+  type?: SchemaType
+  document?: SanityDocument
+  path?: Path
+}
+
+/**
+ * @internal
+ * The base type for all validators in the validation library. Takes in a
+ * `RuleSpec`'s constraint, the value to check, an optional override message,
+ * and the validation context.
+ *
+ * @see Rule.validate from `@sanity/validation/src/Rule`
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type Validator<T = any, Value = any> = (
+  constraint: T,
+  value: Value,
+  message: string | undefined,
+  context: ValidationContext
+) =>
+  | ValidationError[]
+  | ValidationError
+  | string
+  | true
+  | Promise<ValidationError[] | ValidationError | string | true>
+
+/**
+ * @internal
+ * A type helper used to define a group of validators. The type of the
+ * `RuleSpec` constraint is inferred via the key.
+ *
+ * E.g.
+ *
+ * ```ts
+ * const booleanValidators: Validators = {
+ *   ...genericValidator,
+ *
+ *   presence: (v, value, message) => {
+ *     if (v === 'required' && typeof value !== 'boolean') return message || 'Required'
+ *     return true
+ *   },
+ * }
+ * ```
+ */
+export type Validators = Partial<
+  {
+    [P in RuleSpec['flag']]: Validator<
+      ConditionalIndexAccess<Extract<RuleSpec, {flag: P}>, 'constraint'>
+    >
+  }
+>
+
+export interface ValidationErrorOptions {
+  paths?: Path[]
+  children?: ValidationMarker[]
+  operation?: 'AND' | 'OR'
+}
+
+// This follows the same pattern as `RuleClass` and `Rule` above
+export interface ValidationErrorClass {
+  new (message: string, options?: ValidationErrorOptions): ValidationError
+}
+
+export interface ValidationError extends Omit<Error, 'name'> {
```

Comment:
Any benefit from extending Error here? Could we lean on structural typing instead? Maybe worth checking if we rely on `instanceof Error` checks anywhere (I doubt we do, tbh).

And even if we do, it will still work runtime, so not extending Error here could be a good way of "soft deprecating" it I guess

Reply:
@bjoerge i like it! i removed the `extends Error` but i don't think I can remove the class (e.g. in favor of just structural typing/plain objects, because of the `cloneWithMessage` method. we could switch to a factory function but it seems like the class is fine for this use case for now

Reply:
Agreed, class is fine.